python/handleInvalidState.py

```diff
@@ -83,19 +85,49 @@ def generateSafeHeadingDegrees(state, invalidStartObstacle):
     return globalWindDirectionDegrees
 
 
-def checkGoalValidity(state):
-    '''Checks if the goal waypoint of the given state is valid (no obstacle there)
+def checkGoalValidity(state, goalLatlon=None):
+    '''Checks if the goal of the given state is valid (no obstacle there)
 
     Args:
        state (BoatState): State of the sailbot and other boats
+       goalLatlon (latlon): Location to check the validity of. If is None, goal is the current global waypoint
 
     Returns:
        bool True iff there is no obstacle projected to be on the state's goal waypoint
     '''
-    referenceLatlon = state.globalWaypoint
+    if goalLatlon is None:
+        goalLatlon = state.globalWaypoint
+
+    referenceLatlon = goalLatlon
     obstacles = obs.getObstacles(state, referenceLatlon)
     for obstacle in obstacles:
-        goalXY = utils.latlonToXY(state.globalWaypoint, referenceLatlon)
+        goalXY = utils.latlonToXY(goalLatlon, referenceLatlon)
         if not obstacle.isValid(goalXY):
             return False
     return True
+
+
+def moveGlobalWaypointUntilValid(state):
+    '''Moves global waypoint towards sailbot until it is valid
+
+    Args:
+       state (BoatState): State of the sailbot and other boats
+
+    Returns:
+        latlon goalLatlon such that checkGoalValidity(state, newGoal) is true
```

Comment:
Could you add something in the docstring about whether the new waypoint will be further TOWARD or AWAY from sailbot? Eg. will new point move on straight line toward sailbot or away from sailbot. 

Eg. "draws straight line between sailbot and current global waypoint, then moves the goalLatlon along this line TOWARD (or AWAY? tbh unsure) until it is valid.